File: LeetCodeProblemsGo/LeetCode2483MinimumPenaltyForAShop/LeetCode2483MinimumPenaltyForAShop.go
package LeetCode2483MinimumPenaltyForAShop

func bestClosingTime(customers string) int {
	length := len(customers)
	prefixSums := make([]int, length+1)
	suffixSums := make([]int, length+1)
	for i := 1; i <= length; i++ {
		prefixSums[i] = prefixSums[i-1]
		if customers[i-1] == 'N' {
			prefixSums[i]++
		}
	}
	for i := length - 1; i >= 0; i-- {
		suffixSums[i] = suffixSums[i+1]
		if customers[i] == 'Y' {
			suffixSums[i]++
		}
	}
	minIndex := 0
	minPenalty := length
	for i := 0; i <= length; i++ {
		if prefixSums[i]+suffixSums[i] < minPenalty {
			minPenalty = prefixSums[i] + suffixSums[i]
			minIndex = i
		}
	}
	return minIndex
}

File: LeetCodeProblemsCMake/Problems/LeetCode2483MinimumPenaltyForAShop.cpp
//
// Created by saeed on 12/26/25.
//

#include "LeetCode2483MinimumPenaltyForAShop.h"

#include <vector>
using namespace std;

int LeetCode2483MinimumPenaltyForAShop::bestClosingTime(string customers) {
    int length = static_cast<int>(customers.length());
    vector<int> prefixSum(length+1, 0);
    vector<int> suffixSum(length+1,0);
    for (int i=1; i<=length; ++i) {
        prefixSum[i]=prefixSum[i-1]+(customers[i-1]=='N'?1:0);
    }
    for (int i=length-1; i>=0; --i) {
        suffixSum[i]=suffixSum[i+1]+(customers[i]=='Y'?1:0);
    }
    int minPenalty=length;
    int minIndex{-1};
    for (int i=0;i<=length;++i) {
        if ( suffixSum[i]+prefixSum[i]<minPenalty) {
            minPenalty=suffixSum[i]+prefixSum[i];
            minIndex=i;
        }
    }
    return minIndex;
}
